### PR TITLE
Added some visualizations for the HF dataset

### PR DESCRIPTION
**Pull request by htahir1**
## Describe changes
With the ability to embed any Huggingface dataset, this feature adds the visualization the ZenML dashboard, in case one returns a HF dataset.

image

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py [hunk 1]
```diff
@@ -103,3 +133,50 @@ def extract_metadata(
...
+        if isinstance(ds, Dataset):
+            datasets = {"default": ds}
+        elif isinstance(ds, DatasetDict):
+            datasets = ds
+        else:
+            raise ValueError(f"Unsupported type {type(ds)}")
+
+        for name, dataset in datasets.items():
+            # Generate a unique identifier for the dataset
+            dataset_id = extract_dataset_name(
```
Comment (schustmi) [inline src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py:159]: Does this function extract the name or the ID? Either the variable name `dataset_id` or the function name needs to be changed.

File: src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py [hunk 2]
```diff
@@ -33,13 +42,34 @@
 DEFAULT_DATASET_DIR = "hf_datasets"
 
 
+def extract_dataset_name(input_string) -> Optional[str]:
```
Comment (schustmi) [inline src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py:45]: What is `input_string`? The dataset name? The dataset ID? This needs a different name.

File: src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py [hunk 3]
```diff
@@ -103,3 +133,50 @@ def extract_metadata(
...
+                """
+
+                # Save the HTML to a file
+                visualization_path = os.path.join(
+                    self.uri, f"{name}_viewer.html"
+                )
+                with fileio.open(visualization_path, "w") as f:
+                    f.write(html)
+
+                visualizations[visualization_path] = VisualizationType.HTML
```
Comment (schustmi) [inline src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py:180]: The dashboard only works with a single visualization per artifact right now, but that shouldn't stop this PR. Just so you're aware of it.

**Comment by strickvl** [comment]
@coderabbitai review

File: src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py (hunk 2 above)
Comment (strickvl) [inline src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py]:
```suggestion
def extract_dataset_name(input_string: str) -> Optional[str]:
```

Comment (strickvl) [inline src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py]: this function is def asking for a unit test around some of the edge cases here.

Comment (htahir1) [reply src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py]: ill ask claude

Comment (strickvl) [reply src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py]: yeah I got something like:

```python
import pytest
from your_module import extract_repo_url  # Replace 'your_module' with the actual module name

@pytest.mark.parametrize("checksum_str, expected_output", [
    # Normal cases
    ("hf://datasets/nyu-mll/glue@bcdcba79d07bc864c1c254ccfcedcce55bcc9a8c/mrpc/train-00000-of-00001.parquet", "nyu-mll/glue"),
    ("hf://datasets/huggingface/dataset-name@1234567890abcdef1234567890abcdef12345678/subset/file.parquet", "huggingface/dataset-name"),
    
    # Edge cases
    ("", None),  # Empty string
    ("hf://datasets/org/dataset", None),  # Not enough parts
    ("hf://datasets/org/dataset@commit/subset/file.parquet", "org/dataset"),  # Minimal valid input
    ("hf://datasets/org/dataset-with-hyphens@commit/subset/file.parquet", "org/dataset-with-hyphens"),  # Hyphens in dataset name
    ("hf://datasets/org_name/dataset_name@commit/subset/file.parquet", "org_name/dataset_name"),  # Underscores in names
    ("hf://datasets/org.name/dataset.name@commit/subset/file.parquet", "org.name/dataset.name"),  # Dots in names
    
    # Variations
    ("hf://datasets/org/dataset@commit/subset/file-with-hyphens.parquet", "org/dataset"),  # Hyphens in filename
    ("hf://datasets/org/dataset@commit/subset/file_with_underscores.parquet", "org/dataset"),  # Underscores in filename
    ("hf://datasets/org/dataset@commit/subset/file.with.dots.parquet", "org/dataset"),  # Dots in filename
    ("hf://datasets/org/dataset@commit/subset/file.jsonl", "org/dataset"),  # Different file extension
    
    # Invalid inputs (should return None)
    ("not_a_valid_checksum_string", None),
    ("hf:/datasets/org/dataset@commit/subset/file.parquet", None),  # Missing a slash after hf:
    ("https://huggingface.co/datasets/org/dataset", None),  # Different URL format
    (None, None),  # None input
    (123, None),  # Non-string input
])
def test_extract_repo_url(checksum_str, expected_output):
    assert extract_repo_url(checksum_str) == expected_output

def test_extract_repo_url_exception_handling():
    # Test that the function handles exceptions gracefully
    class CustomException(Exception):
        pass
    
    def mock_split_raising_exception(separator):
        raise CustomException("Test exception")
    
    # Monkey patch the split method to raise an exception
    original_split = str.split
    str.split = mock_split_raising_exception
    
    try:
        assert extract_repo_url("hf://datasets/org/dataset@commit/subset/file.parquet") == None
    finally:
        # Restore the original split method
        str.split = original_split
```

Comment (htahir1) [reply src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py]: I did it another way, take a look?

**Review by strickvl — APPROVED** [review]
LGTM, assuming tests pass.

File: src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py [hunk 4]
```diff
@@ -33,13 +42,38 @@
 DEFAULT_DATASET_DIR = "hf_datasets"
 
 
+def extract_repo_url(checksum_str: str) -> Optional[str]:
```
Comment (schustmi) [inline src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py:45]: Again, the output is not a URL.

File: src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py [hunk 5]
```diff
@@ -33,13 +42,38 @@
...
+
+    An example of a checksum_str is:
+    "hf://datasets/nyu-mll/glue@bcdcba79d07bc864c1c254ccfcedcce55bcc9a8c/mrpc/train-00000-of-00001.parquet"
+    and the expected output is "nyu-mll/glue".
+
+    Args:
+        checksum_str: The checksum_str to extract the dataset name from.
+
+    Returns:
+        Optional[str]: The extracted dataset name.
```
Comment (schustmi) [inline src/zenml/integrations/huggingface/materializers/huggingface_datasets_materializer.py:56]:
```suggestion
        The extracted dataset name.
```

This is specified above as a type annotation.

**Review by schustmi — CHANGES_REQUESTED** [review]

**Review by schustmi — APPROVED** [review]
